### PR TITLE
Implement specular AO with screen space bent normals

### PR DESCRIPTION
**Pull request by pixelflinger**
SSAO is optionally augmented to evaluate bent normals, later used for specular AO.

before:
before
after:
after

### CODE REVIEW

**Review by bejado — APPROVED** [review]
Does this require recompiling materials due to the use of an ARRAY texture now?

File: filament/src/materials/ssao/bilateralBlurBentNormals.mat [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+vertex {
+    void postProcessVertex(inout PostProcessVertexInputs postProcess) {
+        postProcess.vertex.xy = postProcess.normalizedUV;
+    }
+}
+
+fragment {
+    #include "ssaoUtils.fs"
+
+    void dummy(){}
```
Comment (romainguy) [inline filament/src/materials/ssao/bilateralBlurBentNormals.mat:56]: Why is this needed again?

File: filament/src/materials/ssao/bilateralBlurBentNormals.mat [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    void postProcess(inout PostProcessInputs postProcess) {
+        highp vec2 uv = variable_vertex.xy; // interpolated at pixel's center
+
+        float ao;
+        highp float depth;
+        highp float sampleDepth;
+        vec3 bn;
+
+        vec3 data = textureLod(materialParams_ssao, vec3(uv, 0.0), 0.0).rgb;
+        ao = data.r;
```
Comment (romainguy) [inline filament/src/materials/ssao/bilateralBlurBentNormals.mat:85]: Why not declare the variables here were you initialize them?

File: filament/src/materials/ssao/bilateralBlurBentNormals.mat [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        depth = unpack(data.gb);
+
+        if (data.g * data.b == 1.0) {
```
Comment (romainguy) [inline filament/src/materials/ssao/bilateralBlurBentNormals.mat:88]: Should it be >= 1.0 just in case?

File: filament/src/materials/ssao/bilateralBlurBentNormals.mat [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            bilateral = weight * bilateralWeight(depth, sampleDepth);
+            totalWeight += bilateral;
+            tapBN(materialParams_ssao, uv - offset, bn);
+            sumAO += ao * bilateral;
+            sumBN += bn * bilateral;
+
+            offset += materialParams.axis;
+        }
+
+        ao = sumAO * (1.0 / totalWeight);
```
Comment (romainguy) [inline filament/src/materials/ssao/bilateralBlurBentNormals.mat:125]: Factor 1/totalWeight in a variable to help the compiler?

File: libs/filabridge/src/SibGenerator.cpp [hunk 5]
```diff
@@ -45,7 +45,7 @@ SamplerInterfaceBlock const& SibGenerator::getPerViewSib(uint8_t variantKey) noe
             .add("froxels",       Type::SAMPLER_2D,         Format::UINT,    Precision::MEDIUM)
             .add("iblDFG",        Type::SAMPLER_2D,         Format::FLOAT,   Precision::MEDIUM)
             .add("iblSpecular",   Type::SAMPLER_CUBEMAP,    Format::FLOAT,   Precision::MEDIUM)
-            .add("ssao",          Type::SAMPLER_2D,         Format::FLOAT,   Precision::MEDIUM)
+            .add("ssao",          Type::SAMPLER_2D_ARRAY,   Format::FLOAT,   Precision::MEDIUM)
```
Comment (romainguy) [inline libs/filabridge/src/SibGenerator.cpp:48]: We need to increase the material version number because of this

**Review by romainguy — APPROVED** [review]

File: filament/src/materials/ssao/bilateralBlurBentNormals.mat (hunk 1 above)
Comment (pixelflinger) [reply filament/src/materials/ssao/bilateralBlurBentNormals.mat:56]: It's just for CLion's glsl editor

File: filament/src/materials/ssao/bilateralBlurBentNormals.mat (hunk 3 above)
Comment (pixelflinger) [reply filament/src/materials/ssao/bilateralBlurBentNormals.mat:88]: No it can't be > 1, since `.g` and `.b` are 0 to 1, also .`gb` encode a depth scaled between 0 and 1 and this encoding exists only because we're storing the depth in 8 bits channels.

File: libs/filabridge/src/SibGenerator.cpp (hunk 5 above)
Comment (pixelflinger) [reply libs/filabridge/src/SibGenerator.cpp:48]: ah good point.
